src/main/java/com/google/devtools/build/lib/actions/usage/ActionInputUsageTracker.java

```diff
@@ -0,0 +1,278 @@
+    }
+
+    /**
+     * Whether action supports input tracking.
+     */
+    public boolean supportsInputTracking(Action action) {
+        return enabled() &&
+                action.getOwner().getLabel() != null &&
+                action.getOwner().getLabel().getRepository().isMain() &&
+                getJDepsOutput(action) != null &&
+                SUPPORTED_DEPENDENCY_TRACKING_MNEMONICS.contains(action.getMnemonic());
+    }
+
+    /**
+     * Whether action supports classes input tracking.
+     */
+    public boolean supportsClassTracking(Action action) {
+        return this.trackerMode == ActionInputUsageTrackerMode.UNUSED_CLASSES &&
+                action.getOwner().getLabel() != null &&
+                action.getOwner().getLabel().getRepository().isMain() &&
+                getJDepsOutput(action) != null &&
+                SUPPORTED_CLASS_TRACKING_MNEMONICS.contains(action.getMnemonic());
+    }
+
+    /**
+     * Refresh internal input tracking info from action .jdeps file.
+     */
+    public void refreshInputTrackingInfo(Action action) {
+        if (!enabled() || !supportsInputTracking(action)) {
+            return;
+        }
+
+        UsageInfo usageInfo = null;
+        try {
+            Artifact jdeps = getJDepsOutput(action);
+            Path output = pathResolver.toPath(jdeps);
+            InputStream input = output.getInputStream();
+            Deps.Dependencies deps = Deps.Dependencies.parseFrom(input);
+            checkState(deps.getRuleLabel().equals(action.getOwner().getLabel().toString()));
+
+            Set<String> usedPaths = deps.getDependencyList().stream()
+                    .filter(d -> d.getKind() == Deps.Dependency.Kind.EXPLICIT || d.getKind() == Deps.Dependency.Kind.IMPLICIT)
+                    .map(d -> d.getPath())
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
+            Set<String> unusedArtifactsPath = action.getInputs().toList().stream()
+                    .filter(ActionInputUsageTracker::canArtifactBeUnused)
+                    .map(d -> d.getExecPathString())
+                    .filter(Predicate.not(usedPaths::contains))
+                    .collect(Collectors.toCollection(LinkedHashSet::new));
+            Map<String, Set<ClassUsageInfo>> usedClassesMap = deps.getDependencyList().stream()
+                    .filter(d -> d.getKind() == Deps.Dependency.Kind.EXPLICIT || d.getKind() == Deps.Dependency.Kind.IMPLICIT)
+                    .collect(Collectors.toMap(
+                            d -> d.getPath(),
+                            d -> d.getUsedClassesList().stream()
+                                    .map(ClassUsageInfo::create)
+                                    .collect(Collectors.toCollection(LinkedHashSet::new))));
+
+            usageInfo = new UsageInfo(unusedArtifactsPath, usedClassesMap);
+        } catch (FileNotFoundException fileNotFoundException) {
+            // Silently ignore, this could be a clean build
```

Comment:
Should this be a `return`?

`ConcurrentHashMap#put` throws an NPE if the key or value is null, so this is causing it to NPE below on clean builds. (It works for me if you do a build with the feature disabled to populate the initial `jdeps`, and then enable it.)

Reply:
@cushon : good catch! We're not hitting this (even with clean build) but this is probably a regression introduced by https://github.com/bazelbuild/bazel/commit/793c62f8ace50db8014c55d6ee58ff1e428f586f. PR is updated now. Thanks!